### ISSUE TITLE
Check if Feature was selected to install.

### ISSUE THREAD

**Issue by jakubpietka** [oleg-shilo/wixsharp#402]
Hi, how can I check if user selected or not feature in tree before installations start? In my case, if user selected feature "IIS application", I want to show one more UI window to allow user specify website port. I know how to show dialog based on condition. I have my UI already created. I don't know how to check right after FeatureTree window in custom action which features was selected to install.
`
customUI.On(NativeDialogs.CustomizeDlg, Buttons.Next, new DialogAction { Name = "DoAction", Value = CheckSelectedFeaturesActionName })`

```
[CustomAction]
public static ActionResult CheckSelectedFeatures(Session session)
{
        System.Windows.Forms.MessageBox.Show(session.Property("ADDLOCAL"));
}
```

Action shows me empty message box.

Thanks.

**Comment by Xaddan** [comment]
Look towards the `IsFeatureEnabled` method.

```C#
[CustomAction]
public static ActionResult CheckSelectedFeatures(Session session)
{
        bool isFeatureNameEnable = session.IsFeatureEnabled("FeatureName");
}
```

**Comment by jakubpietka** [comment]
IsFeatureEnabled returns false, even if feature is selected to install.

```
    [CustomAction]
    public static ActionResult CheckSelectedFeatures(Session session)
    {
        bool isFeatureNameEnable = session.IsFeatureEnabled("Services");

        if (isFeatureNameEnable)
        {
            System.Windows.Forms.MessageBox.Show("Feature Enabled");
        }
        else
        {
            System.Windows.Forms.MessageBox.Show("Feature is not Enabled");
        }

        return ActionResult.Success;
    }
```

Adding action to project:
`project.AddAction(new ManagedAction(CheckSelectedFeaturesActionName) { Id = CheckSelectedFeaturesActionName, Sequence = Sequence.NotInSequence, UsesProperties = "ADDLOCAL=[ADDLOCAL]" });`

This code always shows MessageBox "Feature is not Enabled".

**Comment by oleg-shilo** [comment]
With native UI it can be tricky. You may try to play with the `Sequence = Sequence.NotInSequence`. Fir example try to use `InstallUISequence`.

And I you are using ManagedUI then you can Analise the checkbox state directly.

**Comment by jakubpietka** [comment]
I am using CustomUI. Changing Sequence property to InstallUISequence caused build error: 
`Unresolved reference to symbol 'WixAction:InstallUISequence/InstallInitialize' in section 'Product:{6F330B47-2577-43AD-9095-1861CA25A80C}'`

In my case, I need only default FeatureTree window and show one more window (my own) if one of the feature was selected. Can't I achieve this using Wix#?

I thought that CustomAction after FeatureTree window and Condition after Button.Next will be enough.

```
        var customUI = new CustomUI();
        customUI.On(NativeDialogs.ExitDialog, Buttons.Finish, new CloseDialog() { Order = 9999 });

        customUI.On(NativeDialogs.WelcomeDlg, Buttons.Next, new ShowDialog(NativeDialogs.CustomizeDlg));

        customUI.On(NativeDialogs.CustomizeDlg, Buttons.Next, new DialogAction { Name = "DoAction", Value = CheckSelectedFeaturesActionName }, new ShowDialog(myOwnDialog, Condition.Create("SHOW_DIALOG = YES")));
        customUI.On(NativeDialogs.CustomizeDlg, Buttons.Back, new ShowDialog(NativeDialogs.WelcomeDlg));
        customUI.On(NativeDialogs.CustomizeDlg, Buttons.Cancel, new CloseDialog("MyExit"));

        customUI.On(NativeDialogs.MaintenanceWelcomeDlg, Buttons.Next, new ShowDialog(NativeDialogs.MaintenanceTypeDlg));
        customUI.On(NativeDialogs.MaintenanceTypeDlg, Buttons.Back, new ShowDialog(NativeDialogs.MaintenanceWelcomeDlg));
        customUI.On(NativeDialogs.MaintenanceTypeDlg, Buttons.Repair, new ShowDialog(NativeDialogs.VerifyReadyDlg));
        customUI.On(NativeDialogs.MaintenanceTypeDlg, Buttons.Remove, new ShowDialog(NativeDialogs.VerifyReadyDlg));
```

**Comment by oleg-shilo** [comment]
Unfortunately I cannot provide a comprehensive support for native CustomUI. I can only suggest to look at the generated wxs. The may be a clue there. Quite possibly the check boxes are evaluated by the MSI runtime after you called `CheckSelectedFeaturesActionName` and in result it always has the default value (`false`).

This is what I meant "With native UI it can be tricky". 

With native UI customization Wix# is almost equally unintuitive as raw WiX. That is why I prefer ManagedUI anytime I face a UI customization task.

**Comment by oleg-shilo** [comment]
Yes you can. But probably it is easier if you use VS template "Custom UI":
![image](https://user-images.githubusercontent.com/16729806/41911554-548a8656-7990-11e8-90bd-6c8d64aa1985.png)

It has all standard dialogs with their sources.

**Comment by jakubpietka** [comment]
Ok, I created project with this template and found soultion to skip feature customization dialog if this feature is not selected, but FeatureTree UI actually looks different than NativeDialogs.FeatureTree i was using in custom UI. And the problem is not tree apperance but missing Browse button for Feature ConfigurableDir property. I need to allow user to specify path for each Feature.

**Comment by oleg-shilo** [comment]
ManagedUI is an alternative simplified implementations of the standard native MSI dialogs. In some cases the dialogs user experience was deliberately changed to simplify the overal experience. Thus I do expect that appearance of the specific dialogs can be different comparing to MSI. 

However the project template includes source code for the all dialogs thus the user can adjust the appeasement without any limitations.

As for "Browse" button, I removed it as I saw no practical value for it. Neither for "Disk Usage" button. If you ask me they are those typical _low value but high effort_ features.

But of course you can reinsert the "Browse" button and do whatever native dialog does on clicking it. It will be setting some "magic" MSI property to the selected path value.

**Comment by jakubpietka** [comment]
Main problem is I can't figure it out how modify your dialogs to use my property, it always returns empty string while getting property value. I want to reinsert 'Browse' button so selecting path will update property with feature path.

I started with project (as you can see Dir's ID is MYDIR):
```
 var project = new ManagedProject("MyProduct",
                             new Dir(new Id("MYDIR"),@"%ProgramFiles%\My Company\TEST",
                                 new File("Program.cs")));
```
Then feature:
`var feature = new Feature("TEST", "Description", true, true, "MYDIR");`

And to test it, I just insert this code after user click Next button in my dialog:
```
        void next_Click(object sender, EventArgs e)
        {
            System.Windows.Forms.MessageBox.Show(MsiRuntime.Session.Property("MYDIR"));

            Shell.GoTo(0);
            //Shell.GoNext();
        }
```

but it always shows empty message box.

When I was working with NativeDialogs it worked. What is wrong in this case?

**Comment by oleg-shilo** [comment]
The answer is here: https://github.com/oleg-shilo/wixsharp/wiki/Deployment-scenarios#installation-directory.

I encourage you to drop the `new Id("MYDIR")` and use auto assigned `INSTALLDIR`. Have a look at the  "InstallDirDialog.cs" in your project. It reads and writes the installation directory value.

**Comment by jakubpietka** [comment]
But I have more than one Feature and each of them need different directory, thats why I need 'Browse' button again in FeatureTree. For the same reason INSTALLDIR is not enough, I need more than one directory, each one with specific ID attached to Feature.

**Comment by oleg-shilo** [comment]
There is some confusion...
- **_Dir id_**
In the code sample, you have provided, you were trying to use the the id of the install dir. I guided you through the process of assigning and using the id correctly as your code had a mistake.

- **_Feature dir_**
I am not sure how MSI handles assigning individual directories to the features. That's why I did not give you any advise on that. You will need to dig the MSI/WiX documentation for that. 

However I can guess from the description of the `ConfigurableDirectory` attribute what is required for implementing customization of the feature directory:
```
Specify the Id of a Directory that can be configured by the user at installation time. 
This identifier must be a public property and therefore completely uppercase. 
```
Meaning that the solution may look like this:
```C#
var defaultFeature = new Feature("A", "A files", true, true, "INSTALLDIR"));
var docFeature = new Feature("A", "A files", true, true, "DOCDIR"));

var project =
            new ManagedProject("MyProduct",
                new Dir(@"%ProgramFiles%\My Company\My Product",
                    new File(defaultFeature, @"Files\Bin\MyApp.exe"),
                    new Dir(new Id("DOCDIR"), docFeature, "Documentation",
                        new File(...
```
And in the features dialog collect user selection(s) and assign it to the corresponding feature dir:

```C#
session["INSTALLDIR"] = selectDirDialog.Path
// and for the other one
session["DOCDIR"] = selectDirDialog.Path
```
I am only guessing that as I never tried/needed implementing the behavior you want.
